homeassistant/components/switch/broadlink.py

```diff
@@ -0,0 +1,141 @@
+"""
+Support for Broadlink RM devices.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/switch.broadlink/
+"""
+from datetime import timedelta
+from base64 import b64encode, b64decode
+import binascii
+import logging
+import socket
+import voluptuous as vol
+
+from homeassistant.util.dt import utcnow
+from homeassistant.components.switch import (SwitchDevice, PLATFORM_SCHEMA)
+from homeassistant.const import (CONF_FRIENDLY_NAME, CONF_SWITCHES,
+                                 CONF_COMMAND_OFF, CONF_COMMAND_ON,
+                                 CONF_TIMEOUT, CONF_HOST, CONF_MAC)
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['broadlink==0.2']
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = "broadlink"
+DEFAULT_NAME = 'Broadlink switch'
+DEFAULT_TIMEOUT = 10
+SERVICE_LEARN = "learn_command"
+
+SWITCH_SCHEMA = vol.Schema({
+    vol.Optional(CONF_COMMAND_OFF, default=None): cv.string,
+    vol.Optional(CONF_COMMAND_ON, default=None): cv.string,
+    vol.Optional(CONF_FRIENDLY_NAME, default=DEFAULT_NAME): cv.string,
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_SWITCHES): vol.Schema({cv.slug: SWITCH_SCHEMA}),
+    vol.Required(CONF_HOST): cv.string,
+    vol.Required(CONF_MAC): cv.string,
+    vol.Optional(CONF_TIMEOUT, default=DEFAULT_TIMEOUT): cv.positive_int
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup Broadlink switches."""
+    import broadlink
+    devices = config.get(CONF_SWITCHES, {})
+    switches = []
+    ip_addr = config.get(CONF_HOST)
+    mac_addr = binascii.unhexlify(
+        config.get(CONF_MAC).encode().replace(b':', b''))
+    broadlink_device = broadlink.rm((ip_addr, 80), mac_addr)
+    broadlink_device.timeout = config.get(CONF_TIMEOUT)
+    try:
+        broadlink_device.auth()
+    except socket.timeout:
+        pass
+
+    for object_id, device_config in devices.items():
+        switches.append(
+            BroadlinkRM2Switch(
+                device_config.get(CONF_FRIENDLY_NAME, object_id),
+                device_config.get(CONF_COMMAND_ON),
+                device_config.get(CONF_COMMAND_OFF),
+                broadlink_device
+            )
+        )
+
+    if not switches:
+        _LOGGER.error("No switches added.")
+        return False
+
+    add_devices(switches)
+
+    def _learn_command(call):
+        broadlink_device.enter_learning()
+        start_time = utcnow()
+        while (utcnow() - start_time) < timedelta(seconds=20):
```

Comment:
Will this method take up 100% CPU because it is constant running? 

Makes more sense to use `track_point_in_time` or even betterm make this async and run `check_data` in the executor. Async sleeps take/block no resources.